src/inc/til/bitmap.h

```diff
@@ -0,0 +1,223 @@
+                _pos = _nextPos;
+                _run = til::rectangle{};
+            }
+        }
+    };
+
+    class bitmap
+    {
+    public:
+        using const_iterator = _bitmap_const_iterator;
+
+        bitmap() noexcept :
+            _sz{},
+            _rc{},
+            _bits{},
+            _empty{ true }
+        {
+        }
+
+        bitmap(til::size sz) :
+            _sz(sz),
+            _rc(sz),
+            _bits(sz.area(), false),
+            _empty(true)
+        {
+        }
+
+        const_iterator begin() const
+        {
+            return _bitmap_const_iterator(_bits, _sz, 0);
+        }
+
+        const_iterator end() const
+        {
+            return _bitmap_const_iterator(_bits, _sz, _sz.area());
+        }
+
+        void set(const til::point pt)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(pt));
+
+            til::at(_bits, _rc.index_of(pt)) = true;
+            _empty = false;
+        }
+
+        void reset(const til::point pt)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(pt));
+
+            til::at(_bits, _rc.index_of(pt)) = false;
+        }
+
+        void set(const til::rectangle rc)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(rc));
+
+            for (const auto pt : rc)
+            {
+                set(pt);
+            }
+        }
+
+        void reset(const til::rectangle rc)
+        {
+            THROW_HR_IF(E_INVALIDARG, !_rc.contains(rc));
+
+            for (const auto pt : rc)
+            {
+                reset(pt);
+            }
+        }
+
+        void set_all()
+        {
+            // .clear() then .resize(_size(), true) throws an assert (unsupported operation)
+            // .assign(_size(), true) throws an assert (unsupported operation)
+            set(_rc);
+        }
+
+        void reset_all()
+        {
+            // .clear() then .resize(_size(), false) throws an assert (unsupported operation)
+            // .assign(_size(), false) throws an assert (unsupported operation)
+            reset(_rc);
+            _empty = true;
+        }
+
+        // True if we resized. False if it was the same size as before.
+        bool resize(til::size size)
+        {
+            // Don't resize if it's not different
+            if (_sz != size)
+            {
+                _sz = size;
+                _rc = til::rectangle{ size };
+                // .resize(_size(), true/false) throws an assert (unsupported operation)
+                _bits = std::vector<bool>(_sz.area(), false);
+                _empty = true;
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        constexpr bool empty() const
+        {
+            return _empty;
+        }
+
+    private:
+        bool _empty;
+        til::size _sz;
+        til::rectangle _rc;
+        std::vector<bool> _bits;

```

Comment:
Ok. I personally don't like having `_bits` always be filled with `_sz.area()` falses, then updating this as we go. I now understand why we need a separate `_empty` bool too. I have a few ideas after looking [here](https://en.cppreference.com/w/cpp/container):
1. **make `_bits` a `std::set**
    - Since you can get the index of a given point, when you set a point, add the point's index into the set. When unsetting said point, remove it from the set.
    - This makes us not allocate `area` space on the vector. Resizing will be annoying because you need to do some math to develop a formula that transforms an index from one size to another. But once you figure out the math for resize, you just reuse that. I imagine resizing won't be that common anyways so it's an expense that's worth it.
    - actually, you could even just throw the til::points in directly. If it's in the set, it's set. Otherwise, it's unset.
2. **make `_bits` a `std:: map**
    - It's in the name haha. `key` is the point. `bool` is the val.

Reply:
`std::vector<bool>` is actually special and packs up to 32 or 64 flags into a single processor word. The other types do not have this specialization. We get a massive space advantage for using the specialization.

I have an alternative idea for identifying `none()`, `all()`, `any()`, and `one()` (which are things we need.) Stay tuned.